iati/core/rulesets.py

```diff
@@ -1,45 +1,841 @@
+
+        Args:
+            context_element (Element): An XML Element.
+
+        Returns:
+            bool: Return `True` when one or fewer results are found in the Dataset.
+                  Return `False` when one or more results are found in the Dataset.
+
+        """
+        unique_paths = set(self.paths)
+
+        found_elements = 0
+
+        for path in unique_paths:
+            results = context_element.xpath(path)
+            found_elements += len(results)
+
+        if found_elements > 1:
+            return False
+        return True
+
+
+class RuleRegexMatches(Rule):
+    """Representation of a Rule that checks that the text of the given `paths` must match the `regex` value."""
+
+    def __init__(self, context, case):
+        """Initialise a `regex_matches` Rule.
+
+        Raises:
+            ValueError: When the case does not contain valid regex.
+
+        """
+        self.name = 'regex_matches'
+
+        super(RuleRegexMatches, self).__init__(context, case)
+
+        try:
+            re.compile(self.regex)
+        except sre_constants.error:
+            raise ValueError
+        if self.regex == '':
+            raise ValueError
+
+    def __str__(self):
+        """Return string stating what RuleRegexMatches is checking."""
+        if len(self.paths) == 1:
+            return 'Each `{self.paths[0]}` within each `{self.context}` must match the regular expression `{self.regex}`.'.format(**locals())
+        return 'Each instance of `{0}` within each `{self.context}` must match the regular expression `{self.regex}`.'.format('` and `'.join(self.paths), **locals())
+
+    def _check_against_Rule(self, context_element):
+        """Assert that the text of the given `paths` matches the `regex` value.
+
+        Args:
+            context_element (Element): An XML Element.
+
+        Returns:
+            bool: Return `True` when the given `path` text matches the given regex.
+                  Return `False` when the given `path` text does not match the given `regex`.
+
+        """
+        pattern = re.compile(self.regex)
+
+        for path in self.paths:
+            strings_to_check = self._extract_text_from_element_or_attribute(context_element, path)
+            for string_to_check in strings_to_check:
+                if not pattern.search(string_to_check):
+                    return False
+        return True
+
+
+class RuleRegexNoMatches(Rule):
+    """Representation of a Rule that checks that the text of the given `paths` must not match the `regex` value."""
+
+    def __init__(self, context, case):
+        """Initialise a `regex_no_matches` Rule.
+
+        Raises:
+            ValueError: When the case does not contain valid regex.
+
+        """
+        self.name = 'regex_no_matches'
+
+        super(RuleRegexNoMatches, self).__init__(context, case)
+
+        try:
+            re.compile(self.regex)
+        except sre_constants.error:
+            raise ValueError
+        if self.regex == '':
```

Comment:
This check should possibly be before the try-catch... though maybe not, I'm not entirely sure of the logic my head is making.

Reply:
done